### PR TITLE
Feat(core): add helmet middleware

### PR DESCRIPTION
**Pull request by AkhileshJyotishi**
# Pull Request Guidelines

Our guidelines for submitting a pull request.

## Before submitting a Pull Request, please make sure you have verified the following:

- [x] The commit message follows our guidelines:
  - A good commit message should be two things: meaningful and concise. It should not contain every single detail, describing each changed line—we can see all the changes in Git—but, at the same time, it should say enough to avoid ambiguity.
  - We use Microverse's commit message convention
  - The convention stablish that a commit message has to be in the present tense, imperative and lowercase.
  - Example: `fix typo in README.md`
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Please describe the current behavior that you are modifying, or link to a relevant issue.

Issue Number: #107

## What is the new behavior?

Describe the new behavior or link to a relevant issue.
addHelmet() method added
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications below.

## Other information
This commit adds the 'addHelmet' function to the core module, allowing developers to enhance security by setting various HTTP headers using Helmet middleware. The function accepts optional configuration options for Helmet, providing flexibility in securing web applications.

### CODE REVIEW

File: packages/core/src/middleware/interfaces/helmet.interface.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
+/// <reference types="node" />
```
Comment (rsaz) [inline packages/core/src/middleware/interfaces/helmet.interface.ts:1]: You can remove the directive as the project tsconfig contains node in types already 
```json
"types": ["node", "reflect-metadata", "vitest/globals"]
```

**Review by rsaz — CHANGES_REQUESTED** [review]
You did 90%, and you're in the right path to complete this task. Good job so far. However you're missing the different interfaces `addHelmet({})` could have, for example:

```typescript
// This sets custom options for the
// Content-Security-Policy header.
  helmet({
    contentSecurityPolicy: {
      directives: {
        "script-src": ["'self'", "example.com"],
      },
    },
  })
);
```
image

```typescript
helmet({
    contentSecurityPolicy: false,
    xDownloadOptions: false,
  }),
  ```

Please make sure all interfaces are implemented from helmet, all the configuration you would be able to use on `app.use` you should be able to do in the `addHelmet()` method from your middleware.

Here are the list:

image

Reference: https://www.npmjs.com/package/helmet

**Comment by rsaz** [comment]
** This is how you test your application step by step**

> @rsaz I have implemented the required interface and added it in `middleware-service.ts`. But can you pin point how can I test it?

In order for you test you have to build your library. Here are all the steps prior to the build:

1 - install the cli globally
```bash
npm i @expressots/cli
```
2 - scaffold a project using the cli
```bash
expressots new app-test -t opinionated -p npm
```
3 - in the project dependencies of the app-test remove @expressots/core
```bash
npm uninstall @expressots/core
```
4 - rename your .env.example in to .env only

5 -Now let's build the library with your changes
```bash
npm run build
```
If everything goes right it will generate a `/lib` folder container all lib code including your changes

6 - Let's generate the package to test locally. `npm pack` will generate `.tgz` package in which you can install in your app-test to replace the `@expressots/core` package you removed in the step 3
```bash
npm pack
```
7 - it will be something like: `npm i  expressots-core-2.3.0.tgz`

After done that just run `npm run dev` in your app-test application.
Go in providers/application folder in the `application.provider.ts` mentioned in the issue content right in the top.
And in the section of configuring services you should be able to invoke your middleware for example:

```typescript
protected configureServices(): void {
        this.middleware.addHelmet({}); // and pass your parameters, which represents the interfaces you have created
         this.middleware.addBodyParser();
        
        this.middleware.setErrorHandler();
    }
```
The first test should be:
1 - if you dont have the helmet package installed by the time you execute the application using `npm run dev` you should throw a warning in the console saying, you should install the package for the rate limit middleware. Once you install, this warning should go away then you could use your middleware normally.

2 - to really test your middleware working you should invoke any endpoint already created in the opinionated template. there is a folder user inside of useCases, and inside of this folder you will see several use cases such as create, delete, update and so on.
The controllers of each usecase are the entry points of each request.

Any other doubt you might have you can query the documentation. Happy coding!

**Comment by AkhileshJyotishi** [comment]
I've locally tested the first point you mentioned but the second point isn't clear to me, can you please elaborate?

**Comment by rsaz** [comment]
> I've locally tested the first point you mentioned but the second point isn't clear to me, can you please elaborate?

Invoke the endpoint to test the usage of helmet configuration:

`http://localhost:3000/user/create`

Payload

```json
{
   "name": "user 01",
   "email": "user01@email.com"
}
```

Send a post request and see the headers the configuration for helmet.

**Comment by AkhileshJyotishi** [comment]
@rsaz 
test-video link for reference 
https://www.loom.com/share/69f7c21ec9ab4456b61a4773bd3e47cc

**Review by rsaz — APPROVED** [review]

**Review by rsaz — APPROVED** [review]

**Comment by rsaz** [comment]
> @rsaz test-video link for reference https://www.loom.com/share/69f7c21ec9ab4456b61a4773bd3e47cc

Good job! How was your experience working in the framework?

**Comment by AkhileshJyotishi** [comment]
Thanks !

**Pull request merged**